torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -163,6 +181,16 @@ void ProcessGroupNCCL::WorkNCCL::synchronize() {
       AT_CUDA_CHECK(cudaDeviceSynchronize());
     }
   }
+
+  if (!isSuccess()) {
+    // Set the appropriate exception if found.
+    checkAndSetException();
+
+    // Throw an exception, only if we have a valid exception.
+    if (exception()) {
+      std::rethrow_exception(exception());
+    }
+  }
```

Comment:
Above, `isSuccess` is defined as "there is no exception on the communicator AND the events associated with kernel completion are ready". This means the path here is hit if "there is an exception on the communicator OR the events are not yet ready" (i.e. the kernel has yet to run or is still running). Then it checks for exceptions on the communicator and throws it if there is one. But, if the second part of the conditional is true, then there will be no exception, and the function returns. If after that the communicator still fails, we have a problem. The `wait()` function that the user called has returned and new CUDA kernels will have been queued to run, but there is no way to raise the yet-to-happen exception on a user thread.

Moreover, the watchdog will abort the communicator, and work scheduled after it will run, assuming the operation completed successfully. A worst case scenario here is that an optimizer runs on garbage gradients, the user saves a checkpoint, figures out an error happened on the next iteration, restarts, and restores from a garbage checkpoint.

At the moment, I don't think there is a way we can avoid doing full on synchronization on completion here.

Reply:
If the user really wants to check for all errors, they need to do something like this:

```
work = pg.allreduce(tensors)
pg.barrier()->wait()
if (!work.isSuccess()) {
  // handle error.
}
```

My feeling is that the decision on whether or not to continue other operations without checking if previous operations have succeeded or failed is on the user. If the user decides to use the result of an operation which has failed (probably garbage gradients) without checking its status, its on them.

Another option the user has is to batch a lot of operations together to run asynchronously and then have a single barrier to wait for completion and check for errors.

> Moreover, the watchdog will abort the communicator, and work scheduled after it will run, assuming the operation completed successfully. A worst case scenario here is that an optimizer runs on garbage gradients, the user saves a checkpoint, figures out an error happened on the next iteration, restarts, and restores from a garbage checkpoint.

In this case, doesn't the user need to have a barrier before running the optimizer (to ensure the previous operation completed?). If so, the user should check for the status of the previous operation after the barrier.

Reply:
Just to clarify, my comment here was w.r.t. the use of `isSuccess` inside `wait`, not the definition of `isSuccess` by itself. The current use of `isSuccess` allows for `wait` to return without actually checking for completion of the operation.

> My feeling is that the decision on whether or not to continue other operations without checking if previous operations have succeeded or failed is on the user. If the user decides to use the result of an operation which has failed (probably garbage gradients) without checking its status, its on them.

This is counter-intuitive w.r.t. the current behavior. Today, if you call `wait()`, either the subsequent operations will execute on the real result of the operation, or your processes will hang indefinitely. Adding an in-between mode where the subsequent operations execute, but they execute on garbage, will lead to unexpected results for users. Therefore I strongly feel that `wait` must synchronize. 

> Another option the user has is to batch a lot of operations together to run asynchronously and then have a single barrier to wait for completion and check for errors.

This is what DDP does as well -- first it kicks off all the work it needs to do, asynchronously, and then when it's done with autograd, it waits on all of them in order. In practice, most of them will have already completed in the background, so the synchronization won't involve actually waiting for something to complete.

> In this case, doesn't the user need to have a barrier before running the optimizer (to ensure the previous operation completed?). If so, the user should check for the status of the previous operation after the barrier.

This shouldn't be a separate operation, or we open ourselves up to the unexpected behavior. It should be all part of `wait` in my opinion. There is also no need for a barrier, if `wait` is indeed synchronous.

Reply:
>This is counter-intuitive w.r.t. the current behavior. Today, if you call wait(), either the subsequent operations will execute on the real result of the operation, or your processes will hang indefinitely. Adding an in-between mode where the subsequent operations execute, but they execute on garbage, will lead to unexpected results for users. Therefore I strongly feel that wait must synchronize.

Will update the PR to make `wait()` synchronous. Although, doesn't this change the behavior for current users? The operation was non-blocking and now we are making it blocking. Is there any way to inform users about this change apart from adding documentation to the method?

Reply:
Would I be correct if I assume the above questions are already resolved after introducing `NCCL_BLOCKING_WAIT ` env var?